### PR TITLE
Set DistModel state_dict keys to structure_names

### PR DESCRIPTION
**Pull request by pangengzheng**
### PR types

Others
### PR changes

Others
### Description

card-78318
Set DistModel state_dict keys to structure_names, thus support dynamic graph save static load and static save dynamic graph load

### CODE REVIEW

File: python/paddle/distributed/auto_parallel/api.py [hunk 1]
```diff
@@ -1257,6 +1263,15 @@ def state_dict(self, mode="all"):
...
         dist_state_dict = self._build_distributed_state_dict(local_state_dict)
+        mapping_names = [
+            self._parameter_to_structured_name[k]
+            if k in self._parameter_to_structured_name
+            else k
+            for k in dist_state_dict.keys()
+        ]
+        dist_state_dict = dict(
+            zip(mapping_names, list(dist_state_dict.values()))
+        )
```
Comment (zhiqiu) [inline python/paddle/distributed/auto_parallel/api.py:1274]: why k may not in _parameter_to_structured_name?

File: python/paddle/distributed/checkpoint/load_state_dict.py [hunk 2]
```diff
@@ -37,14 +37,36 @@ class ReadItem:
...
 
 
-def get_rank_to_files(path, state_dict, process_group, use_dist):
+PATH_TO_CHECKPOINT_FILES: Dict[str, Tuple[list, list]] = {}
+
+
+def get_checkpoint_files(path, use_cache=True):
+    global PATH_TO_CHECKPOINT_FILES
+    if use_cache and path in PATH_TO_CHECKPOINT_FILES:
+        return PATH_TO_CHECKPOINT_FILES[path]
```
Comment (zhiqiu) [inline python/paddle/distributed/checkpoint/load_state_dict.py:46]: why need cache?

File: python/paddle/distributed/auto_parallel/api.py (hunk 1 above)
Comment (pangengzheng) [reply python/paddle/distributed/auto_parallel/api.py:1274]: 因为program的state_dict返回的可能有优化器参数，优化器state_dict的key不在这个_parameter_to_structured_name里面

File: python/paddle/distributed/checkpoint/load_state_dict.py (hunk 2 above)
Comment (pangengzheng) [reply python/paddle/distributed/checkpoint/load_state_dict.py:46]: 为了减少系统调用os.listdir，系统调用相对比较耗时

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**